lib/scholar/linear/logistic_regression.ex

```diff
@@ -85,127 +83,92 @@ defmodule Scholar.Linear.LogisticRegression do
-
-    diff = Nx.reshape(logit - y_t, {m})
-
-    coeff_diff = Nx.dot(x_t, diff) / m
-
-    bias_diff = Nx.sum(diff) / m
-
-    new_coeff = coeff - coeff_diff * learning_rate
-    new_bias = bias - bias_diff * learning_rate
-
-    {new_coeff, new_bias}
+  defnp loss_fn(coeff, bias, xs, ys) do
+    probs = softmax(Nx.dot(xs, coeff) + bias)
+    cross_entropy(probs, ys)
   end
 
-  # Gradient descent for multinomial regression
-
-  defnp update_coefficients_multinomial(x, x_t, one_hot, coeff, learning_rate) do
-    {m, _n} = x.shape
-
-    dot_prod = Nx.dot(x, coeff)
-
-    prob = softmax(dot_prod)
+  defnp grad_loss(coeff, bias, xs, ys) do
+    grad({coeff, bias}, fn {coeff, bias} -> loss_fn(coeff, bias, xs, ys) end)
+  end
 
-    diff = Nx.dot(x_t, prob - one_hot) / m
+  # Normalized softmax
 
-    coeff - diff * learning_rate
+  defn softmax(t) do
+    normalized_exp = (t - Nx.reduce_max(t, axes: [-1], keep_axes: true)) |> Nx.exp()
+    normalized_exp / Nx.sum(normalized_exp, axes: [-1], keep_axes: true)
   end
 
   @doc """
   Makes predictions with the given model on inputs `x`.
 
   ## Examples
 
-      > x = Nx.tensor([[1.0, 2.0], [3.0, 2.0], [4.0, 7.0]])
-      > y = Nx.tensor([4.0, 3.0, -1.0])
-      > model = Scholar.Linear.LogisticRegression.fit(x, y, num_classes: 2)
-      > Scholar.Linear.LogisticRegression.predict(model, Nx.tensor([-3.0, 5.0]))
-      Nx.tensor(
-        0
-      )
+      iex> x = Nx.tensor([[1.0, 2.0], [3.0, 2.0], [4.0, 7.0]])
+      iex> y = Nx.tensor([1, 0, 1])
+      iex> model = Scholar.Linear.LogisticRegression.fit(x, y, num_classes: 2)
+      iex> Scholar.Linear.LogisticRegression.predict(model, Nx.tensor([[-3.0, 5.0]]))
+      #Nx.Tensor<
+        s64[1]
+        [1]
+      >
   """
-  defn predict(%__MODULE__{mode: mode} = model, x) do
-    case mode do
-      :binary -> predict_binary(model, x)
-      :multinomial -> predict_multinomial(model, x)
-    end
+  defn predict(%__MODULE__{coefficients: coeff, bias: bias}, x) do
+    inter = Nx.dot(x, Nx.transpose(coeff)) + bias
+    Nx.argmax(inter, axis: 1)
   end
 
-  defnp predict_binary(%__MODULE__{coefficients: coeff, bias: bias}, x) do
-    logit = 1 / (Nx.exp(-(Nx.dot(coeff, Nx.transpose(x)) + bias)) + 1)
-    logit > 0.5
-  end
+  @doc """
+  Calculates probabilities of predictions with the given model on inputs `x`.
+
+  ## Examples
 
-  defnp predict_multinomial(%__MODULE__{coefficients: coeff}, x) do
-    dot_prod = Nx.dot(x, coeff)
-    Nx.argmax(dot_prod, axis: 1)
+      iex> x = Nx.tensor([[1.0, 2.0], [3.0, 2.0], [4.0, 7.0]])
+      iex> y = Nx.tensor([1, 0, 1])
+      iex> model = Scholar.Linear.LogisticRegression.fit(x, y, num_classes: 2)
+      iex> Scholar.Linear.LogisticRegression.predict_proba(model, Nx.tensor([[-3.0, 5.0]]))
+      #Nx.Tensor<
+        f32[1][2]
+        [
+          [1.650987258017267e-7, 0.9999998807907104]
+        ]
+      >
+  """
+  defn predict_proba(%__MODULE__{coefficients: coeff, bias: bias}, x) do
```

Comment:
Do we use the "proba" abbreviation anywhere else? Maybe we shouldn't abbreviate it?